### ISSUE TITLE
runsc inside of default docker seccomp policy

### ISSUE THREAD

**Issue by prattmic** [google/gvisor#4371]
@scanlime on Twitter is trying to [run runsc inside a Docker container](https://twitter.com/scanlime/status/1310630296694398977) with the standard seccomp [policy](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) enabled. This is similar to rootless mode (#311), but a little bit more strict.

The immediate issue is that we exec into empty namespaces, which the profile does not allow. It is not clear if there would be more issues if that were resolved, though I didn't see any glaring issues comparing our seccomp filters to Docker's.

It's also not clear if the defense-in-depth features we'd have to disable to make this work would make it a bad idea. But in general, it is very reasonable to want to run a sandbox as a subprocess in an existing container.

cc @fvoznika @nlacasse

**Comment by scanlime** [comment]
Hi, prattmic invited me here to explain a bit further. Thanks for opening the issue. It's still in the experimental stages but I'm working on a media server that will be doing transcodes in a locked down environment, and I've been investigating which approaches are possible to adopt without adding a portability burden. For wide use i would really want the resulting project to run in a docker container, and adding extra privileges to that container for the promise of better sandboxing within the container is really a non-starter. So that has me looking at either forking gvisor and collapsing it to the bare minimum really, or writing a tiny seccomp and ldpreload thing just for this purpose.

**Comment by fvoznika** [comment]
That's an interesting use case. gVisor setups up many layers of defense around the sandbox to reduce access to host in case the sandbox is compromised. You can find more details in the [Containing a Real Vulnerability](https://gvisor.dev/blog/2020/09/18/containing-a-real-vulnerability/#defense-in-depth) blog post. Configuring some of these layers require `CAP_SYS_ADMIN` that are not available to Docker containers unless `docker run --privileged` flag is used. This is unintuitive, but in the same way that Docker requires to be running as root to create containers, gVisor needs root to protect the sandbox as well.

So you can remove some of the layers that require high privilege (e.g. namespaces, pivot_root), similar to the way `--rootless` does. You just need to be aware that in the end the sandbox will not be as protected. You need to decide whether this protection is good enough for your use case. On the positive side, the most important security layer is [seccomp](https://cs.opensource.google/gvisor/gvisor/+/master:runsc/boot/filter/config.go), which you can still use.

Another consideration is that the gofer requires [these capabilities](https://cs.opensource.google/gvisor/gvisor/+/master:runsc/cmd/gofer.go;l=40;drc=da07e38f7c986dca568466beea8a0b7db60f09db) to function correctly. Otherwise, some operations will not be allowed, like creating files with different owners.

A few other options to consider are:
1. Run everything inside gVisor: this works if you don't need to isolate media server from the code that would be running inside gVisor
1. Use different containers: run the media server in one container and untrusted workload in runsc. You can connect then using container network. This would require giving the media server access to the Docker unix socket to run containers.

**Comment by avagin** [comment]
https://github.com/avagin/gvisor/commit/1924e9e258e0ee687db120ab3d70f36d93209f12

This is the POC patch which adds the --unprivileged flag.  With this flag, the gVisor can be executed in a default docker container with limitations that Fabricio described in the previous comment.

```
$ docker run -it --rm -v /tmp/runsc:/mnt alpine /bin/sh
/ # 
/ # /mnt/runsc --rootless --unprivileged --network none do echo 'Hello World!'
Hello World!
```

If the unprivileged flag is specified, Sentry and Gofer processes are running in the current set of namespace. By other words, we remove one level of isolation. But in your case, a docker container provides you this extra level, so I think your use-case can be still valid.

**Comment by scanlime** [comment]
> This is the POC patch which adds the --unprivileged flag. With this flag, the gVisor can be executed in a default docker container with limitations that Fabricio described in the previous comment.

This is interesting! How do filesystem and PID isolation work in this case? It looks from my untrained reading that this would allow access to send signals to other processes owned by the same user, or open files accessible to the current user. Does the emulated kernel provide that level of isolation?

**Comment by prattmic** [comment]
Yes, the userspace kernel still provides isolation, because it is emulating the OS based on the provided configuration. These capabilities are providing defense-in-depth in the event that the kernel is compromised.

In the case of signals, all thread IDs within the sandbox are entirely internal to the userspace kernel, with no relation to the host. Signal syscalls sent by the sandboxed application can only target other threads in the sandbox (implementing a signal syscall in the userspace kernel may not even send a host signal at all). In fact, using a PID namespace is really a third layer of defense, since the userspace kernel [can't send signals to other processes anyways](https://github.com/google/gvisor/blob/master/runsc/boot/filter/config.go#L288).

Similarly, the userspace kernel can't directly open host files. That is mediated by another process called the gofer. The gofer won't grant access to files not allowed by the configuration, but were it to be compromised, the mount namespace containing only the configured files provides an additional layer of protection.

**Comment by teng1** [comment]
This is interesting, slightly different use-case I was also trying to use runsc inside a container in an attempt to get a dockerised (proprietary) application to work on a hardened K8S platform, the application requires some privileged capabilities to work however the container platform drops these for security reasons. 

The idea was to use gvisor in a pod to (in a crude sense) pick these capabilities back up as a compatibility layer for the app. Would such a thing be feasible?

**Comment by fvoznika** [comment]
It's not a generic solution that will work with all containers, but it may work in your case depending on what the container does at runtime. In gVisor, all file system operations are handled by an external file proxy, called Gofer, that is isolated from the sandbox for security purposes. The gofer [requires capabilities](https://cs.opensource.google/gvisor/gvisor/+/master:runsc/cmd/gofer.go;drc=da07e38f7c986dca568466beea8a0b7db60f09db;l=40) to function correctly. For example, when the container creates a file running as an user that exists inside the sandbox, the gofer requires `CAP_CHOWN` to [change the file owner](https://cs.opensource.google/gvisor/gvisor/+/master:runsc/fsgofer/fsgofer.go;drc=da07e38f7c986dca568466beea8a0b7db60f09db;l=445) after creation.

**Comment by paulfitz** [comment]
Is there any possibility the `--unprivileged` flag from @avagin's POC could be added to mainstream gvisor? It comes in handy from time to time, for the kind of use-case outlined at start of thread. I could imagine perhaps fearing that people would turn it on without understanding the trade-offs.

**Comment by LarsSven** [comment]
For my project I must also run gVisor inside a docker container for integration-testing purposes. It is not possible to do this outside of a docker container, as there are other requirements the environment has like specific file system mounting and Linux-specific tools, while the test environment must be triggered locally from systems that don't have gVisor installed, file systems mounted, or are even Linux.

Is there any way the unprivileged flag could somehow be merged or updated?

**Comment by dsagal** [comment]
Chiming in with another request to merge in this flag. It's needed for some use cases! Any reason not to merge it in?

We've been maintaining a fork with this flag, and it's a pain to maintain a fork... If the concern is that people may use `--unprivileged` without understanding tradeoffs, a scarier name would be fine. Perhaps `--omit-privileged-isolation`?

**Comment by jonathanperret** [comment]
Hello! We've been working to keep @avagin's patch up to date with recent versions of gVisor because we have use cases as mentioned above that require running  `runsc` inside a Docker or Kubernetes container, in environments where setting gVisor up as a runtime is not feasible.

Ideally this would get integrated into gVisor `master` so we wouldn't have to keep a fork. We're willing to contribute some work to that effect, and I see this issue is now flagged `help wanted`, is that open to external contributions?

If so, a few pointers would be helpful — I've read `CONTRIBUTING.md` but it has not made entirely clear what level of testing is expected before submitting a PR. Are all the tests run by `bazel test ...` expected to pass? It seems those include a fair number of tests that require specific environments, at the very least running on bare metal — I've been working in an EC2 VM so far and a lot of stuff isn't even starting. The `CONTRIBUTING.md` document refers to Presubmit checks accessible with membership to the `gvisor-dev` list but I've joined the list and am not clear about the next step.

Thanks for any help!

**Comment by EtiennePerot** [comment]
@jonathanperret 

> We're willing to contribute some work to that effect, and I see this issue is now flagged `help wanted`, is that open to external contributions?

Yes, although perhaps not in this form. IMO the biggest problem is way the flag is specified; it should be changed to better reflect its security implications, and it should still try to do as much as whatever the host policy allows it to get away with. For example, the flag could instead be specified as something like like `--self-sandboxing-policy=STRICT|BEST_EFFORT`, where the default is `STRICT` and models current gVisor behavior, and `BEST_EFFORT` reflects "try as you can to sandbox yourself but it's OK if you can't" behavior. Note that isn't quite the same as what this patch is doing (it's unconditionally bypassing all self-sandboxing measures, rather than trying if they can work).

The main other issue is needing additional integration tests to be maintainable, i.e. extend https://github.com/google/gvisor/blob/master/test/e2e/runtime_in_docker_test.go to cover these cases.

> Are all the tests run by `bazel test ...` expected to pass?

Please use the `make test TARGETS=...` idiom instead, but other than that yes. Feel free to file separate issues if they don't. For the test I mentioned, you need to run it on a machine with Docker installed, with user permission to run Docker containers, where the default runtime is `runc` (Docker's default), and where the integration test image is loaded. I patched it (#13674) to make it easier to run.

```
# Load host Docker with basic test images (needed only once)
$ make load-basic

# Run test:
$ make test TARGETS=//test/e2e:runtime_in_docker_test
```
